src/main/scala/util/HeterogeneousBag.scala

```diff
@@ -6,12 +6,34 @@ import Chisel._
 import chisel3.Record
 import scala.collection.immutable.ListMap
 
-final case class HeterogeneousBag[T <: Data](elts: Seq[T]) extends Record with collection.IndexedSeq[T] {
+final case class HeterogeneousBag[T <: Data](elts: Seq[T], eltNames: Option[Seq[String]] = None)
+    extends Record with collection.IndexedSeq[T] {
+
+  eltNames.foreach{ n =>
+    require(elts.size == n.size,
+    s"If eltNames are specified, they must be specified for all elts " +
+      s" (eltNames.size = ${n.size} but elts.size = ${elts.size})")
+    require(n.distinct.size == n.size, "If eltNames are specified they must all be specified & unique")
+  }
+
+  private def nameFromIndex(i: Int): String = eltNames.map(_(i)).getOrElse(i.toString)
+
+  private val nameToElts: Map[String, T] = ListMap(elts.zipWithIndex.map { case (elt, i) =>
+    val name = nameFromIndex(i)
+    name -> elt
+  }:_*)
+
   def apply(x: Int) = elts(x)
+
+  def apply(s: String) = nameToElts(s)
+
   def length = elts.length
 
-  val elements = ListMap(elts.zipWithIndex.map { case (n,i) => (i.toString, n) }:_*)
-  override def cloneType: this.type = (new HeterogeneousBag(elts.map(_.chiselCloneType))).asInstanceOf[this.type]
+  val elements = ListMap(elts.zipWithIndex.map { case (elt, i) =>
```

Comment:
are `elements` and `nameToElts` the same thing?

Reply:
hmm, maybe! I was just replacing the string code but seems like you're right.

Reply:
this was all reverted